src/components/menu-item/submenu-controller.ts

```diff
@@ -0,0 +1,108 @@
+import '../popup/popup.js';
+
+import { createRef, ref, Ref } from 'lit/directives/ref.js';
+import { HasSlotController } from '../../internal/slot.js';
+import { html } from 'lit';
+import { LocalizeController } from '../../utilities/localize.js';
+import SlMenuItem from './menu-item.js';
+import SlPopup from '../popup/popup.js';
+import type { ReactiveController, ReactiveControllerHost } from 'lit';
+
+/** A reactive controller to manage the registration of event listeners for submenus. */
+export class SubmenuController implements ReactiveController {
+  private host: ReactiveControllerHost & SlMenuItem;
+  private popupRef: Ref<SlPopup> = createRef();
+  private mouseOutTimer: number = -1;
+  private isActive: boolean = false;
+
+  private readonly hasSlotController: HasSlotController;
+  private readonly localize: LocalizeController;
+
+  constructor(
+    host: ReactiveControllerHost & SlMenuItem,
+    hasSlotController: HasSlotController,
+    localize: LocalizeController
+  ) {
+    (this.host = host).addController(this);
+    this.hasSlotController = hasSlotController;
+    this.localize = localize;
+  }
+
+  hostConnected() {
+    if (this.hasSlotController.test('submenu') && !this.host.disabled) {
+      this.addListeners();
+    }
+  }
+
+  hostDisconnected() {
+    this.removeListeners();
+  }
+
+  hostUpdated() {
+    if (this.hasSlotController.test('submenu') && !this.host.disabled) {
+      this.addListeners();
+    } else {
+      this.removeListeners();
+    }
+  }
+
+  private addListeners() {
+    if (!this.isActive) {
+      this.host.addEventListener('mouseover', this.handleMouseOver);
+      this.host.addEventListener('mouseout', this.handleMouseOut);
+      this.isActive = true;
+    }
+  }
+
+  private removeListeners() {
+    if (this.isActive) {
+      this.host.removeEventListener('mouseover', this.handleMouseOver);
+      this.host.removeEventListener('mouseout', this.handleMouseOut);
+      this.isActive = false;
+    }
+  }
+
+  private handleMouseOver = () => {
+    clearTimeout(this.mouseOutTimer);
+    if (this.hasSlotController.test('submenu') && this.popupRef.value) {
+      this.popupRef.value.active = true;
+    }
+  };
+
+  private handleMouseOut = () => {
+    if (this.popupRef.value && this.popupRef.value.active) {
+      this.mouseOutTimer = window.setTimeout(() => {
+        if (this.popupRef.value && this.popupRef.value.active) {
+          this.popupRef.value.active = false;
+          //this.isMouseOver = false;
+        }
+      }, 100);
+    }
+  };
+
+  renderSubmenu() {
+    // Always render the slot. Conditionally render the outer sl-popup.
+
+    if (!this.isActive) {
+      return html` <slot name="submenu" hidden></slot> `;
+    }
+
+    const isLtr = this.localize.dir() === 'ltr';
+    return html`
+      <style>
+        ::part(popup) {
+          z-index: var(--sl-z-index-dropdown);
+        }
+      </style>
+      <sl-popup
+        ${ref(this.popupRef)}
```

Comment:
This is a bit of a nit, but can we use `@query` instead? This is the only thing importing ref, but many things import query so that will be one less module we have to worry about bringing in.

Reply:
At this location, `this` is a `ReactiveController`, and the query directive wants `this` to be a `ReactiveElement` (or at least supporting `.querySelector()`). I don't believe the `@query` directive supports changing the `renderRoot` / querySelector root, but I'm happy to be shown otherwise.

Access to this element could be wrapped in a call to `.querySelector()` (which is what `@query` does anyway). ?

I'm slightly surprised there's no non-decorator equivalent I could use, i.e. in the manner this uses `createRef()`; maybe I just don't know where it is.

I'm happy to change it.

`@query` source [here](https://github.com/lit/lit/blob/c134604f178e36444261d83eabe9e578c1ed90c4/packages/reactive-element/src/decorators/query.ts#L42)

(Intuitively, refs _should_ have better performance, right?)

Reply:
🤔 I think the ref / queries may be unnecessary. If it were me, I'd probably just write the value of `isActive` into the template.


```js
  private handleMouseOver = () => {
    clearTimeout(this.mouseOutTimer);
    if (this.hasSlotController.test('submenu') && this.isActive === false) {
      this.isActive = true;
      this.host.requestUpdate()
    }
  }
  
   render () {
      <sl-popup
        ?active=${this.isActive}
        placement=${isLtr ? 'right-start' : 'left-start'}
        anchor="anchor"
        flip
        strategy="fixed"
      >
        <slot name="submenu"></slot>
      </sl-popup>
   }
```

The `requestUpdate()` is important because itll force the re-render since ReactiveControllers dont have reactive properties.